test/functional/plugin/lsp/snippet_spec.lua

```diff
@@ -148,5 +148,47 @@ describe('vim.lsp._snippet', function()
     }, parse('${1|\\,,\\||}'))
   end)
 
+  it('should parse format', function()
+    eq({
+      type = snippet.NodeType.SNIPPET,
+      children = {
+        {
+          type = snippet.NodeType.VARIABLE,
+          name = 'VAR',
+          transform = {
+            type = snippet.NodeType.TRANSFORM,
+            pattern = 'regex',
+            format = {
+              {
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+                modifier = 'upcase',
+              },
+              {
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+                if_text = 'if_text',
+              },
+              {
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+				else_text = 'else_text',
+              },
+              {
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+				else_text = 'else_text',
+				if_text = 'if_text',
+              },
+              {
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+				else_text = 'else_text',
+              },
```

Comment:
Sorry. Actually, I don't know the meaning of `':+' if` like format.

In my eyes, this parsed results does not contain the `+` or `-` or `?` context. Is this expected?

(Anyway, Thank you for your fix!)

Reply:
```lua
S.format = P.any(
  P.map(P.seq(S.dollar, S.int), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[2],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, S.slash, P.any(
    P.token('upcase'),
    P.token('downcase'),
    P.token('capitalize'),
    P.token('camelcase'),
    P.token('pascalcase')
  ), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      modifier = values[6],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.question, P.take_until({ ':' }, { '\\' }), S.colon, P.take_until({ '}' }, { '\\' })),  S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      if_text = values[5][2].esc,
      else_text = (values[5][4] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.plus, P.opt(P.take_until({ '}' }, { '\\' }))), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      if_text = (values[5][2] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.minus, P.opt(P.take_until({ '}' }, { '\\' }))), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      else_text = (values[5][2] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.opt(P.take_until({ '}' }, { '\\' })) ,  S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      else_text = (values[5] or {}).esc,
    }, Node)
  end)
)
```

In my local environment, the above parser can pass your tests without modifying the parser combinators. 👍🏼

Reply:
Thank you for taking a look :)

> OK. VSCodes doesn't have context of + and -. It's just a kind of if/else value.

Yeah, just `if_text` for existing/nonemtpy capture and `else_text` otherwise, at least that's what I got when playing around with it in vscode.

>In my local environment, the above parser can pass your tests without modifying the parser combinators. 👍🏼

Right, the parser combinator-modifications are necessary for handling empty strings, I'll add a few tests for those (and get this ready for review in general :sweat_smile:)

Reply:
```lua
parse('${VAR/regex/${1:/upcase}${1:+if_text}${1:-else_text}${1:?if_text:else_text}${1:else_text}${1:+}${1:-}${1:?:}${1:}/}')
```

My local code was passed the above test case without parser combinator-modifications :)
I think you should use `P.opt` instead of `prevent_empty`.

Reply:
Note: If you want to treat the omitted value as `''` (empty string). You can convert `else_text = (values[5] or {}).esc or ''`

Reply:
> ```lua
> parse('${VAR/regex/${1:/upcase}${1:+if_text}${1:-else_text}${1:?if_text:else_text}${1:else_text}${1:+}${1:-}${1:?:}${1:}/}')
> ```
> 
> My local code was passed the above test case without parser combinator-modifications :) I think you should use `P.opt` instead of `prevent_empty`.

Ahhh sorry, I didn't take that close a look at it.
`opt` is indeed much better👍👍